packages/edge-config/src/utils/tracing.ts

```diff
@@ -1,14 +1,82 @@
-import type { TracerProvider, Tracer } from '@opentelemetry/api';
+import type { TracerProvider, Tracer, Attributes } from '@opentelemetry/api';
 import { name as pkgName, version } from '../../package.json';
 
-let tracer: Tracer | undefined;
+let tracer: TracerProvider | undefined;
 
 export function setTracerProvider(tracerProvider: TracerProvider): void {
-  tracer = tracerProvider.getTracer(pkgName, version);
+  tracer = tracerProvider;
 }
 
-export function getTracer(): Tracer | undefined {
-  return tracer;
+function getTracer(): Tracer | undefined {
+  return tracer?.getTracer(pkgName, version);
+}
+
+export function trace<F extends (...args: any) => any>(
+  fn: F,
+  options: {
+    name: string;
+    tags?: Record<string, string | number | boolean>;
+    tagSuccess?: (
+      result: ReturnType<F> extends PromiseLike<infer U> ? U : ReturnType<F>,
+    ) => Attributes;
+    tagError?: (error: Error) => Attributes;
+  } = {
+    name: fn.name,
+  },
+): F {
+  const traced = function (this: unknown) {
+    const args = arguments as unknown as unknown[];
+    const that = this;
+
+    const tracer = getTracer();
+    if (!tracer) return fn.apply(that, args);
+
+    return tracer.startActiveSpan(options.name, (span) => {
+      if (options.tags) span.setAttributes(options.tags);
+
+      try {
+        const result = fn.apply(that, args);
+
+        if (result instanceof Promise) {
```

Comment:
Do you have a full control of `result` here? There are still a lot of `PromiseLike` instead of promises. To borrow the code from Next.js to test on whether something is a promise:

```
const isPromise = <T>(p: any): p is Promise<T> => {
  return p !== null && typeof p === 'object' && typeof p.then === 'function'
}
```

Reply:
We should, but I guess there never is a guarantee. Will use the same check instead then.